modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java

```diff
@@ -552,4 +554,138 @@ public static String decodeBase64(String receiver) {
         // O(N) or faster depending on implementation
         return result.toArray(new String[0]);
     }
+
+
+    // getByPath public API
+    /**
+     * Access values in nested containers with a dot separated path.  Path elements are treated
+     * as strings for Maps and integers for Lists.
+     * Throws 'IllegalArgumentException' if path does not exist or a path element for a List is not an integer.
+     */
+    public static <E> Object getByPath(List<E> receiver, String path) {
+        return getByPath(receiver, splitPath(path), 0, getByPathMissing(path));
+    }
+
+    /**
+     * Access values in nested containers with a dot separated path.  Path elements are treated
+     * as strings for Maps and integers for Lists.
+     * Throws 'IllegalArgumentException' if path does not exist or a path element for a List is not an integer.
+     */
+    public static <K,V> Object getByPath(Map<K,V> receiver, String path) {
+        return getByPath(receiver, splitPath(path), 0, getByPathMissing(path));
+    }
+
+    /**
+     * Same as 'getByPath(List, String)' but returns defaultValue if path does not exist.
+     * Throws 'IllegalArgumentException' if there is a non-container type at a non-terminal path element.
+     */
+    public static <E> Object getByPath(List<E> receiver, String path, Object defaultValue) {
+        return getByPath(receiver, splitPath(path), 0, () -> defaultValue);
+    }
+
+    /**
+     * Same as 'getByPath(Map, String)' but returns defaultValue if path does not exist.
```

Comment:
I think the main logic you describe needs to be moved up to these methods. There will be no javadocs generated for private methods.

Also, you can refer to other methods with `{@link #methodName(signature...)}`. So you can have most of the javadocs on one of the public methods, then link to that and describe the differences in the others.

Reply:
Moved all docs to top level method and `{@link..}` to it from variants.